agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/telemetry/TelemetryChannel.java

```diff
@@ -364,19 +368,23 @@ private Consumer<Throwable> errorHandler(
       String instrumentationKey, Consumer<Boolean> onFailure, OperationLogger operationLogger) {
 
     return error -> {
-      if (isStatsbeat
-          && (error instanceof UnknownHostException
-              || error instanceof UnresolvedAddressException)) {
-        // when sending a Statsbeat request and server returns an UnknownHostException, it's
-        // likely that it's using AMPLS. In that case, we use the kill-switch to turn off Statsbeat.
+      if (isStatsbeat && statsbeatErrorCount.getAndIncrement() >= MAX_STATSBEAT_ERROR_COUNT) {
+        // when sending a Statsbeat request and server returns an Exception 3 times in a row, it's
+        // likely that it's using AMPLS or other private endpoints. In that case, we use the
+        // kill-switch to turn off Statsbeat.
+        // TODO need to figure out a way to detect AMPL or we can let the new ingestion service to
+        // handle this case for us when it becomes available.
+        statsbeatErrorCount.set(0);
```

Comment:
statsbeatErrorCount should be reset when there is no error, here you should update isStatsbeat flag I believe so this code is never triggered again

Reply:
reset should happen after max is reached. after that this block of code shouldn't happen since statsbeat has been shutting down. isStatsbeat is sent for every statsbeat request. does it make sense?  

getAndIncrement will return the previous value and then increment it. when it reaches 3, reset and then shutdown. reset or not doesn't matter since no more statsbeat going out.

Reply:
If you don't reset the retry counter on success if you have 3 errors even if they are not consecutive, you will stop Statsbeat

Reply:
I think also we only care if it's the very first 3 requests that error

Reply:
Updated.